### DISCUSSION TITLE
Added: LDM Notes for 07/01/2020

### DISCUSSION THREAD

**Discussion by 333fred** [dotnet/csharplang#3636] [LDM Notes]
https://github.com/dotnet/csharplang/blob/master/meetings/2020/LDM-2020-07-01.md

Agenda:
1. Non-defaultable struct types and parameterless struct constructors
2. Confirming unspeakable Clone method implications

Note: This will be the last Wednesday LDM for a while. Despite coronavirus, we're going to proceed with our normal summer slowdown, and hold meetings only on Mondays for the next couple of months to encourage people to take some time off. So the notes aren't missing, they just don't exist :).

**Comment by Joe4evr** [comment]
> One thing that is clear from discussion is that non-defaultable struct types will need to have some standardized form of checking whether they are the default instance. `ImmutableArray<T>`, for example, has an `IsDefault` property, as do most of the existing struct types in Roslyn that cannot be used when `default`.

I would like to throw in [my suggestion of one or two attributes that would inform the flow-analysis](https://github.com/dotnet/csharplang/issues/146#issuecomment-604738843). Even if a more standard form of default checking gets implemented, having these to assist for APIs that already exist and are used as such today will help a lot.

**Comment by ufcpp** [comment]
Can't `struct`s participate in `with` expressions in C# 9.0 milestone? `struct`'s memberwise copy doesn't need a `Clone` method.

**Comment by 333fred** [comment]
> Can't `struct`s participate in `with` expressions in C# 9.0 milestone? `struct`'s memberwise copy doesn't need a `Clone` method.

The Clone things we talked about were specific to record types only. C# 9 will not have struct records.

**Comment by ufcpp** [comment]
`struct` can have init-only properties in C# 9.0. `with` needs only memberwise copy and init-only properties. Nevertheless, won't `with` expressions for value types be implemented until struct records?

**Comment by 333fred** [comment]
> `struct` can have init-only properties in C# 9.0. `with` needs only memberwise copy and init-only properties. Nevertheless, won't `with` expressions for value types be implemented until struct records?

I'm unsure what you're asking for here. We didn't discuss anything about struct records or structs in with expressions in this meeting.

**Comment by ufcpp** [comment]
Are `with` expressions completely part of records and not considered an orthogonal feature like init-only properties? There is a demand for non-destructive mutation for immutable value types but it's not discussed before struct records?

**Comment by ufcpp** [comment]
`record` is something like a reference type that have `struct` semantics. The unspeakable `Clone` methods discussed in this LDM are very similar to `struct`'s implicit copy.

**Comment by ufcpp** [comment]
> Therefore whatever future method we add that will allow a regular class type to participate in a with expression will likely have to emit this method as well.

Does `struct` also need a user-overridable Clone method in the future?

**Comment by CyrusNajmabadi** [comment]
> Does struct also need a user-overridable Clone method in the future?

It has not been designed yet.

**Comment by ufcpp** [comment]
OK. I understand this is out of scope for this LDM but my concern is that regular structs will be deprecated when record structs are added, like anonymous methods after lambda expressions.

**Comment by orthoxerox** [comment]
@333fred Could you please spell out the name of the month in the LDM notes issue names? It will save everyone from DD/MM countries at least 15 seconds of confusion per LDM.

**Comment by CyrusNajmabadi** [comment]
> OK. I understand this is out of scope for this LDM but my concern is that regular structs will be deprecated when record structs 

There is no way that regular structs would be deprecated.

**Comment by orthoxerox** [comment]
Some older community discussions for structs with custom invalid values:

- https://github.com/dotnet/roslyn/issues/4347
- https://github.com/dotnet/roslyn/issues/15108
- https://github.com/dotnet/csharplang/issues/147
- https://github.com/dotnet/csharplang/issues/167
- https://github.com/dotnet/csharplang/issues/494

**Comment by 333fred** [comment]
> @333fred Could you please spell out the name of the month in the LDM notes issue names? It will save everyone from DD/MM countries at least 15 seconds of confusion per LDM.

I'll keep that in mind. It is fully spelled out in the notes themselves.
